backend/core/src/shard/connector.rs

```diff
@@ -0,0 +1,156 @@
+            if Instant::now().duration_since(act.hb) > CLIENT_TIMEOUT {
+                // stop actor
+                ctx.close(Some(CloseReason {
+                    code: ws::CloseCode::Abnormal,
+                    description: Some("Missed heartbeat".into()),
+                }));
+                ctx.stop();
+            }
+        });
+    }
+
+    fn handle_message(&mut self, msg: ShardMessage, ctx: &mut <Self as Actor>::Context) {
+        let ShardMessage { conn_id, payload } = msg;
+
+        // TODO: get `NodeId` for `ShardConnId` and proxy payload to `self.chain`.
+    }
+
+    fn start_frame(&mut self, bytes: &[u8]) {
+        if !self.contbuf.is_empty() {
+            log::error!("Unused continuation buffer");
+            self.contbuf.clear();
+        }
+        self.continue_frame(bytes);
+    }
+
+    fn continue_frame(&mut self, bytes: &[u8]) {
+        if self.contbuf.len() + bytes.len() <= CONT_BUF_LIMIT {
+            self.contbuf.extend_from_slice(&bytes);
+        } else {
+            log::error!("Continuation buffer overflow");
+            self.contbuf = BytesMut::new();
+        }
+    }
+
+    fn finish_frame(&mut self) -> Bytes {
+        mem::replace(&mut self.contbuf, BytesMut::new()).freeze()
+    }
+}
+
+impl StreamHandler<Result<ws::Message, ws::ProtocolError>> for ShardConnector {
+    fn handle(&mut self, msg: Result<ws::Message, ws::ProtocolError>, ctx: &mut Self::Context) {
+        self.hb = Instant::now();
+
+        let data = match msg {
+            Ok(ws::Message::Ping(msg)) => {
+                ctx.pong(&msg);
+                return;
+            }
+            Ok(ws::Message::Pong(_)) => return,
+            Ok(ws::Message::Text(text)) => text.into_bytes(),
+            Ok(ws::Message::Binary(data)) => data,
+            Ok(ws::Message::Close(reason)) => {
+                ctx.close(reason);
+                ctx.stop();
+                return;
+            }
+            Ok(ws::Message::Nop) => return,
+            Ok(ws::Message::Continuation(cont)) => match cont {
+                Item::FirstText(bytes) | Item::FirstBinary(bytes) => {
+                    self.start_frame(&bytes);
+                    return;
+                }
+                Item::Continue(bytes) => {
+                    self.continue_frame(&bytes);
+                    return;
+                }
+                Item::Last(bytes) => {
+                    self.continue_frame(&bytes);
+                    self.finish_frame()
+                }
+            },
+            Err(error) => {
+                log::error!("{:?}", error);
+                ctx.stop();
+                return;
+            }
+        };
+
+        match bincode::options().deserialize(&data) {
+            Ok(msg) => self.handle_message(msg, ctx),
+            #[cfg(debug)]
+            Err(err) => {
+                log::warn!("Failed to parse shard message: {}", err,)
+            }
+            #[cfg(not(debug))]
+            Err(_) => (),
```

Comment:
This is a bit dangerous no? I imagine you don't want to spam the logs in production, but isn't it a bit risky to not know about deserialization issues? Can't we let devops take care of log ingestion (and eventually filter out unneeded logs) for us?